File: cpp/133/Clone_Graph.cpp
/**
 * Definition for undirected graph.
 * struct UndirectedGraphNode {
 *     int label;
 *     vector<UndirectedGraphNode *> neighbors;
 *     UndirectedGraphNode(int x) : label(x) {};
 * };
 */
class Solution {
public:
    UndirectedGraphNode *cloneGraph(UndirectedGraphNode *node) {
        if(!node) return NULL;
        return clone(node);
    }
    //深度优先搜索
    UndirectedGraphNode *clone(UndirectedGraphNode *node){
        if(clonedNodesMap.find(node)!=clonedNodesMap.end())
            return clonedNodesMap[node];
        UndirectedGraphNode *newNode = new UndirectedGraphNode(node->label);
        clonedNodesMap[node] = newNode;
        for(UndirectedGraphNode *neighbor:node->neighbors){
            newNode->neighbors.push_back(clone(neighbor));
        }
        return newNode;
    }

private:
    map<UndirectedGraphNode*,UndirectedGraphNode*> clonedNodesMap;
};

File: python/133/Clone_Graph.py
"""
# Definition for a Node.
class Node(object):
    def __init__(self, val, neighbors):
        self.val = val
        self.neighbors = neighbors
"""
class Solution(object):
    def cloneGraph(self, node):
        """
        :type node: Node
        :rtype: Node
        """
        d = dict()
        def dfs(v):
            n = Node(v.val, [])
            d[v] = n
            for x in v.neighbors:
                if not x in d:
                    d[x] = dfs(x)
                n.neighbors.append(d[x])
            return n
        return dfs(node)
